src/twisted/mail/smtp.py

```diff
@@ -1651,8 +1651,9 @@ def listExtensions(self):
         for c, v in self.extensions().items():
             if v is not None:
                 if v:
+                    bv = [x.encode() if isinstance(x, str) else x for x in v]
```

Comment:
The [docstring of `extensions()`](https://github.com/twisted/twisted/blob/d415545aca4697ad033f5bc135d9afa8fe470309/src/twisted/mail/smtp.py#L1628) states that it returns "a `dict` with `bytes` keys and a value of either `None` or a `list` of `bytes`." Therefore I think the problem isn't that `listExtensions()` is not accepting `str` values, but that it is given `str` values.

You wrote on IRC that you'd like to allow strings as dictionary keys, but that represents an interface change, which means you'd have far more work to do:
- convince the Twisted developers that the change is an improvement (technically I am a Twisted developer, but I'd rather leave this decision to devs who have more history with the SMTP code)
- update the docstring of `extensions()`
- check that all other callers of `extensions()` can handle `str` keys and if not, update them too
- think about what happens if the same extension is defined twice, once with its name as `bytes` and once with its name as `str`

The alternative solution that you mentioned on IRC, changing the keys in the [`emailserver.tac` example](https://github.com/twisted/twisted/blob/d415545aca4697ad033f5bc135d9afa8fe470309/docs/mail/examples/emailserver.tac#L65) to `bytes`, would be the easier and safer way to fix this, in my opinion.

Reply:
Sounds like you're right! The only argument I have for dict keys being strings is that it makes them [play better with the json serializer](https://docs.python.org/3/library/json.html#json.dumps) which I assume is not a consideration here.